File: apps/web/src/lib/api/media.ts
import { apiClient } from './client';

export interface MediaAsset {
    id: string;
    draft_id: string;
    type: 'image' | 'carousel_slide' | 'cover_image';
    storage_url: string;
    alt_text: string | null;
    slide_index: number | null;
    visual_template_id: string | null;
    generation_prompt: string | null;
    metadata_json: Record<string, any>;
    created_at: string;
}

export interface ImageRenderRequest {
    draft_id: string;
    template_id: string;
    variables: Record<string, any>;
    brand_overrides?: Record<string, string> | null;
}

export interface ImageGenerateRequest {
    draft_id: string;
    prompt?: string | null;
    style?: string;
    width?: number;
    height?: number;
}

export interface ImageUploadResponse {
    media_asset: MediaAsset;
    message: string;
}

export const mediaApi = {
    /**
     * Render an image from a visual template with variables
     */
    renderTemplate: async (data: ImageRenderRequest): Promise<ImageUploadResponse> => {
        const response = await apiClient.post('/media/render', data);
        return response.data;
    },

    /**
     * Generate an AI image using Gemini Imagen
     */
    generateImage: async (data: ImageGenerateRequest): Promise<ImageUploadResponse> => {
        const response = await apiClient.post('/media/generate', data);
        return response.data;
    },

    /**
     * Upload a user-provided image
     */
    uploadImage: async (
        draftId: string,
        file: File,
        altText?: string,
        slideIndex?: number,
    ): Promise<ImageUploadResponse> => {
        const formData = new FormData();
        formData.append('draft_id', draftId);
        formData.append('file', file);
        if (altText) formData.append('alt_text', altText);
        if (slideIndex !== undefined) formData.append('slide_index', String(slideIndex));

        const response = await apiClient.post('/media/upload', formData, {
            headers: { 'Content-Type': 'multipart/form-data' },
        });
        return response.data;
    },

    /**
     * List all media assets for a draft
     */
    listDraftMedia: async (draftId: string): Promise<MediaAsset[]> => {
        const response = await apiClient.get(`/media/draft/${draftId}`);
        return response.data;
    },

    /**
     * Delete a media asset
     */
    deleteMedia: async (assetId: string): Promise<void> => {
        await apiClient.delete(`/media/${assetId}`);
    },
};

File: apps/api/app/schemas/media.py
"""Media asset schemas."""
from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class MediaAssetResponse(BaseModel):
    """Schema for media asset response."""

    id: UUID
    draft_id: UUID
    type: str
    storage_url: str
    alt_text: Optional[str] = None
    slide_index: Optional[int] = None
    visual_template_id: Optional[UUID] = None
    generation_prompt: Optional[str] = None
    metadata_json: dict = Field(default_factory=dict)
    created_at: datetime

    class Config:
        from_attributes = True


class ImageRenderRequest(BaseModel):
    """Request to render an image from a visual template."""

    draft_id: UUID
    template_id: UUID
    variables: dict = Field(..., description="Template variable values")
    brand_overrides: Optional[dict] = Field(
        default=None,
        description="Optional brand customization: {primary_color, font_family, ...}",
    )


class ImageGenerateRequest(BaseModel):
    """Request to generate an AI image."""

    draft_id: UUID
    prompt: Optional[str] = Field(
        default=None,
        description="Image prompt. Auto-generated from draft content if blank.",
    )
    style: str = Field(
        default="professional",
        description="Style preset: professional, creative, minimal, bold, abstract",
    )
    width: int = Field(default=1080, ge=256, le=2048)
    height: int = Field(default=1080, ge=256, le=2048)


class ImageUploadResponse(BaseModel):
    """Response after uploading/generating an image."""

    media_asset: MediaAssetResponse
    message: str = "Image created successfully"
